lib/matplotlib/__init__.py

```diff
@@ -424,138 +422,157 @@ def wrapper(*args, **kwargs):
     return wrapper
 
 
+_ExecInfo = namedtuple("_ExecInfo", "executable version")
+
+
+@functools.lru_cache()
+def get_executable_info(name):
+    """
+    Get the version of some executable that Matplotlib optionally depends on.
+
+    .. warning:
+       The list of executables that this function supports is set according to
+       Matplotlib's internal needs, and may change without notice.
+
+    Parameters
+    ----------
+    name : str
+        The executable to query.  The following values are currently supported:
+        "dvipng", "gs", "inkscape", "pdftops".  This list is subject to change
+        without notice.
+
+    Returns
+    -------
+    If the executable is found, a namedtuple with fields ``executable`` (`str`)
+    and ``version`` (`distutils.version.LooseVersion`, or ``None`` if the
+    version cannot be determined); ``None`` if the executable is not found.
```

Comment:
Also `None` if the version smaller than the minimal required version. Which can crash later usage patterns like `str(get_executable_info("dvipng").version)`.

I think, this function does too much. The version check should not be part of getting the version.

Reply:
Good points.  I'll try to come up with a better API.

Reply:
What about
1) `require_executable(exec_name)` which returns the correct name to invoke the executable if it exists and its version is recent enough, and throws some exception with a reasonable error message if either the executable doesn't exist, or its version is too old;
2) `list_executables()` which lists install status and version info for all "relevant" executables (mostly to get debugging info; I actually don't particularly think that's needed but I promised that to get #9571 in :-) or we could even just have a `mpl.get_debug_info()` similar to `pandas.show_versions()` which just dumps all kinds of whatever we think may be relevant in bug reports: versions of dependencies, versions of IPython/Jupyter, version of external executables, etc.)

Note that while 1) may seem a bit overkill just for ghostscript, there's actually at least another tool that would benefit from this, which is convert (imagemagick) in the animation framework), which currently also involves a sophisticated dance on Windows.